File: app/services/prometeo_api.py
"""
Prometeo API Client - Open Banking para Argentina
Flujo: usuario ingresa credenciales bancarias → Prometeo las valida → devuelve session key → fetch de cuentas/movimientos
Docs: https://docs.prometeoapi.com
"""
import os
import httpx
from datetime import datetime, timedelta
from typing import Dict, List, Optional

SANDBOX_URL = "https://banking.sandbox.prometeoapi.net"
PRODUCTION_URL = "https://banking.prometeoapi.net"


class PrometeoClient:
    def __init__(self):
        self.api_key = os.getenv("PROMETEO_API_KEY", "")
        self.env = os.getenv("PROMETEO_ENV", "sandbox")
        self.base_url = SANDBOX_URL if self.env == "sandbox" else PRODUCTION_URL
        if not self.api_key:
            raise ValueError("PROMETEO_API_KEY requerido en .env")

    def _headers(self) -> Dict[str, str]:
        return {
            "X-API-Key": self.api_key,
            "Content-Type": "application/json",
        }

    async def list_providers(self) -> List[Dict]:
        """Listar bancos disponibles en Prometeo"""
        try:
            async with httpx.AsyncClient() as client:
                r = await client.get(
                    f"{self.base_url}/provider/",
                    headers=self._headers(),
                    timeout=30.0,
                )
                if r.status_code == 200:
                    data = r.json()
                    providers = data if isinstance(data, list) else data.get("providers", [])
                    # Filtrar solo Argentina o devolver todos
                    return [
                        {"code": p.get("code") or p.get("name"), "name": p.get("name", p.get("code", "Unknown"))}
                        for p in providers
                        if isinstance(p, dict)
                    ]
                print(f"Prometeo providers error {r.status_code}: {r.text[:200]}")
                return []
        except Exception as e:
            print(f"Error listando providers Prometeo: {e}")
            return []

    async def login(self, provider: str, username: str, password: str, doc_type: str = "C") -> Optional[Dict]:
        """
        Login a un banco vía Prometeo.
        Retorna dict con session_key y status.
        Posibles status: logged_in, select_client, wrong_credentials, error
        """
        try:
            async with httpx.AsyncClient() as client:
                r = await client.post(
                    f"{self.base_url}/login/",
                    headers=self._headers(),
                    json={
                        "provider": provider,
                        "username": username,
                        "password": password,
                        "type": doc_type,
                    },
                    timeout=30.0,
                )
                if r.status_code == 200:
                    data = r.json()
                    return {
                        "session_key": data.get("key"),
                        "status": data.get("status"),
                        "message": data.get("message", ""),
                    }
                print(f"Prometeo login error {r.status_code}: {r.text[:200]}")
                return {"status": "error", "message": f"Error {r.status_code}"}
        except Exception as e:
            print(f"Error en login Prometeo: {e}")
            return {"status": "error", "message": str(e)}

    async def get_accounts(self, session_key: str) -> List[Dict]:
        """Obtener cuentas usando session key"""
        try:
            async with httpx.AsyncClient() as client:
                r = await client.get(
                    f"{self.base_url}/account/",
                    headers=self._headers(),
                    params={"key": session_key},
                    timeout=30.0,
                )
                if r.status_code == 200:
                    data = r.json()
                    accounts = data if isinstance(data, list) else data.get("accounts", [])
                    normalized = []
                    for acc in accounts:
                        normalized.append({
                            "account_id": acc.get("id") or acc.get("number"),
                            "name": acc.get("name", "Cuenta"),
                            "number": acc.get("number", ""),
                            "currency": acc.get("currency", "ARS"),
                            "balance": float(acc.get("balance", 0)),
                            "type": acc.get("type", ""),
                        })
                    return normalized
                print(f"Prometeo accounts error {r.status_code}: {r.text[:200]}")
                return []
        except Exception as e:
            print(f"Error obteniendo cuentas Prometeo: {e}")
            return []

    async def get_movements(self, session_key: str, account_number: str, currency: str = "ARS", days: int = 90) -> List[Dict]:
        """Obtener movimientos de una cuenta"""
        start = (datetime.now() - timedelta(days=days)).strftime("%d/%m/%Y")
        end = datetime.now().strftime("%d/%m/%Y")
        try:
            async with httpx.AsyncClient() as client:
                r = await client.get(
                    f"{self.base_url}/movement/",
                    headers=self._headers(),
                    params={
                        "key": session_key,
                        "account_number": account_number,
                        "currency": currency,
                        "date_start": start,
                        "date_end": end,
                    },
                    timeout=30.0,
                )
                if r.status_code == 200:
                    data = r.json()
                    movements = data if isinstance(data, list) else data.get("movements", [])
                    normalized = []
                    for mov in movements:
                        amount = float(mov.get("debit", 0) or 0) - float(mov.get("credit", 0) or 0)
                        if amount == 0:
                            amount = float(mov.get("amount", 0) or 0)
                        date_raw = mov.get("date", "")
                        try:
                            # Prometeo devuelve fechas como DD/MM/YYYY
                            if "/" in date_raw:
                                d, m, y = date_raw.split("/")
                                date_iso = f"{y}-{m.zfill(2)}-{d.zfill(2)}"
                            else:
                                date_iso = date_raw[:10]
                        except Exception:
                            date_iso = datetime.now().date().isoformat()

                        normalized.append({
                            "transaction_id": mov.get("id", f"prom_{account_number}_{date_iso}_{amount}"),
                            "account_id": account_number,
                            "amount": amount,
                            "date": date_iso,
                            "name": mov.get("description", "Transacción"),
                            "merchant_name": mov.get("description", ""),
                            "personal_finance_category": {"primary": "OTHER", "detailed": "OTHER"},
                            "pending": False,
                            "iso_currency_code": currency,
                        })
                    return normalized
                print(f"Prometeo movements error {r.status_code}: {r.text[:200]}")
                return []
        except Exception as e:
            print(f"Error obteniendo movimientos Prometeo: {e}")
            return []


# Lazy global
_instance: Optional[PrometeoClient] = None


def get_prometeo_client() -> Optional[PrometeoClient]:
    global _instance
    if _instance is None and os.getenv("PROMETEO_API_KEY"):
        try:
            _instance = PrometeoClient()
        except ValueError:
            return None
    return _instance

File: services/transactions-service/src/services/prometeoApi.js
/** Prometeo API Client — Open Banking para Argentina. Docs: docs.prometeoapi.com */
import { settings } from "@guita-coach/shared";
const SANDBOX_URL = "https://banking.sandbox.prometeoapi.net";
const PRODUCTION_URL = "https://banking.prometeoapi.net";
class PrometeoClient {
  constructor() {
    this.baseUrl = settings.prometeoEnv === "sandbox" ? SANDBOX_URL : PRODUCTION_URL;
    if (!settings.prometeoApiKey) throw new Error("PROMETEO_API_KEY requerido en .env");
  }
  headers() {
    return {
      "X-API-Key": settings.prometeoApiKey,
      "content-type": "application/json"
    };
  }
  async listProviders() {
    try {
      const r = await fetch(`${this.baseUrl}/provider/`, {
        headers: this.headers()
      });
      if (!r.ok) return [];
      const data = await r.json();
      const providers = Array.isArray(data) ? data : data.providers || [];
      return providers.filter(p => p && typeof p === "object").map(p => ({
        code: p.code || p.name,
        name: p.name || p.code || "Unknown"
      }));
    } catch {
      return [];
    }
  }
  async login(provider, username, password, docType = "C") {
    try {
      const r = await fetch(`${this.baseUrl}/login/`, {
        method: "POST",
        headers: this.headers(),
        body: JSON.stringify({
          provider,
          username,
          password,
          type: docType
        })
      });
      if (!r.ok) return {
        status: "error",
        message: `Error ${r.status}`
      };
      const data = await r.json();
      return {
        session_key: data.key,
        status: data.status,
        message: data.message || ""
      };
    } catch (e) {
      return {
        status: "error",
        message: String(e)
      };
    }
  }
  async getAccounts(sessionKey) {
    try {
      const url = new URL(`${this.baseUrl}/account/`);
      url.searchParams.set("key", sessionKey);
      const r = await fetch(url, {
        headers: this.headers()
      });
      if (!r.ok) return [];
      const data = await r.json();
      const accounts = Array.isArray(data) ? data : data.accounts || [];
      return accounts.map(acc => ({
        account_id: acc.id || acc.number,
        name: acc.name || "Cuenta",
        number: acc.number || "",
        currency: acc.currency || "ARS",
        balance: Number(acc.balance || 0),
        type: acc.type || ""
      }));
    } catch {
      return [];
    }
  }
  async getMovements(sessionKey, accountNumber, currency = "ARS", days = 90) {
    const start = fmtDate(new Date(Date.now() - days * 86400000));
    const end = fmtDate(new Date());
    try {
      const url = new URL(`${this.baseUrl}/movement/`);
      url.search = new URLSearchParams({
        key: sessionKey,
        account_number: accountNumber,
        currency,
        date_start: start,
        date_end: end
      }).toString();
      const r = await fetch(url, {
        headers: this.headers()
      });
      if (!r.ok) return [];
      const data = await r.json();
      const movements = Array.isArray(data) ? data : data.movements || [];
      return movements.map(mov => {
        let amount = Number(mov.debit || 0) - Number(mov.credit || 0);
        if (amount === 0) amount = Number(mov.amount || 0);
        const dateRaw = String(mov.date || "");
        let dateIso;
        if (dateRaw.includes("/")) {
          const [d, m, y] = dateRaw.split("/");
          dateIso = `${y}-${m.padStart(2, "0")}-${d.padStart(2, "0")}`;
        } else {
          dateIso = dateRaw.slice(0, 10) || new Date().toISOString().slice(0, 10);
        }
        return {
          transaction_id: mov.id || `prom_${accountNumber}_${dateIso}_${amount}`,
          account_id: accountNumber,
          amount,
          date: dateIso,
          name: mov.description || "Transacción",
          merchant_name: mov.description || "",
          personal_finance_category: {
            primary: "OTHER",
            detailed: "OTHER"
          },
          pending: false,
          iso_currency_code: currency
        };
      });
    } catch {
      return [];
    }
  }
}
function fmtDate(d) {
  return `${String(d.getDate()).padStart(2, "0")}/${String(d.getMonth() + 1).padStart(2, "0")}/${d.getFullYear()}`;
}
let instance = null;
export function getPrometeoClient() {
  if (!instance && settings.prometeoApiKey) {
    try {
      instance = new PrometeoClient();
    } catch {
      return null;
    }
  }
  return instance;
}
